### DISCUSSION TITLE
Automatic Tool Start by Reservation

### DISCUSSION THREAD

**Discussion by Defenestrar** [usnistgov/NEMO#244] [Ideas]
First of all, I love this project (and NIST).  I've talked to a number of administrators who run NEMO and most of us would have a use case for tool usage starting (and ending) automatically based on reservation (likely a tool specific option).

Each situation is different, but for me I have a couple of instruments that need to be continuously powered (i.e. power interlocks don't work), aren't networkable (so software interlocks don't work), and in a bad zone for wireless/wi-fi (problematic for self-reported tool usage).  My personal case is a need to track/bill based on reserved time (not exact tool usage) and so an option to automatically start and stop tool usage based on the reservation would be great.

I suppose another way to do it would be to allow billing/reports based on reservations, but that might be more complicated to implement due to the systems already built around handling missed reservations.

**Comment by rptmat57** [comment]
Hello and thank you!
I understand your solution, but I feel like if this is going to be done, it should be done "correctly" in the sense that if reservations are being used to charge, then those should be charged, rather than "faking" a tool usage that didn't actually happen in NEMO.

This would keep the data "correct" in NEMO.

Implementation-wise, this would mean:

1. creating a new "reservation" type of charge
2. it would be returned from the billing API and everywhere else (my usage, project billing, etc.) when it's enabled for the tool. 
3. it would be mutually exclusive with missed reservation threshold (add validation for that)
4. regular usage event might need to be disabled?

if you don't mind me asking, what's your facility? and who else did you talk to (just curious)? you can email me directly at mathieu.rampant@nist.gov

**Reply by jgraith** [reply]
Hey Mat,
we are using NEMO not for billing but mainly to track tool usage times etc. 
We also use it globally which causes some difficulties that could be resolved by a similiar approach to the OG request. 
Would it be feasible to auto-disengage after the reservation ends?
We would like to have overnight measurements and people from other timezones can start engaging with the tool again. 

Is there such a function yet and we did not see it?
If not, would it be feasible to implement?
Thanks in advance.

**Reply by rptmat57** [reply]
Hello,
That function does not exist currently. However, in [NEMO-CE](https://gitlab.com/nemo-community/nemo-ce) there is a function allowing users who have a reservation to "take over" the usage of a tool, i.e. end the session for the previous user and start a new one for them. This could be an option.

Otherwise, there is no plan yet to add such a feature in NEMO.
